include/stream-client/impl/logger.ipp

```diff
@@ -0,0 +1,119 @@
+#pragma once
+
+#include <iomanip>
+#include <iostream>
+#include <mutex>
+#include <sstream>
+
+namespace stream_client {
+
+namespace detail {
+
+inline const log_interface* logger_instance_impl(const log_interface* new_logger = nullptr) {
+    static const log_interface* glob_instance = nullptr;
+    if (new_logger) {
+        glob_instance = new_logger;
+    }
+    return glob_instance;
+}
+
+inline logger_message_func logger_func_impl(logger_message_func msg_func = nullptr) {
+    static logger_message_func glob_func = nullptr;
```

Comment:
- Both `glob_instance` and `glob_func` could reference the same instance
- `set_logger_instance`/`set_logger_func` should have same name. They have the same purpose and from the argument, it is clear what operates on.
- We need LOG_* macros, not functions. They help to avoid args evaluation if the level is not enough
- All public free-functions and std_streams_logger should be covered with tests

I'm talking about smth like this
```cpp
#pragma once

#include <chrono>
#include <memory>
#include <mutex>
#include <stdexcept>
#include <string>

/**
 * Internal macro.
 * Cuts filename from the __FILE__.
 */
#ifndef FNAME
#define FNAME (strrchr("/" __FILE__, '/') + 1)
#endif

/**
 * Internal macro.
 * Used to not evaluate LOG_...() arguments if the current level is not enough.
 */
#ifndef STREAM_LOG_CALL
#define STREAM_LOG_CALL(level, ...)                                                                   \
    do {                                                                                              \
        const auto logger = detail::logger_instance();                                                \
        if (logger && logger->get_level() <= level) {                                                 \
            logger->message(level, std::string(FNAME) + ":" + std::to_string(__LINE__), __VA_ARGS__); \
        }                                                                                             \
    } while (0)
#endif

/**
 * Logs the message if the current log level >= to the ERROR level.
 *
 * @note Doesn't construct the message if the current level is less than the ERROR level
 */
#ifndef STREAM_LOG_ERROR
#define STREAM_LOG_ERROR(...) STREAM_LOG_CALL(log_level::error, __VA_ARGS__)
#endif

/**
 * Logs the message if the current log level >= to the WARN level.
 *
 * @note Doesn't construct the message if the current level is less than the WARN level
 */
#ifndef STREAM_LOG_WARN
#define STREAM_LOG_WARN(...) STREAM_LOG_CALL(log_level::warning, __VA_ARGS__)
#endif

/**
 * Logs the message if the current log level >= to the INFO level.
 *
 * @note Doesn't construct the message if the current level is less than the INFO level
 */
#ifndef STREAM_LOG_INFO
#define STREAM_LOG_INFO(...) STREAM_LOG_CALL(log_level::info, __VA_ARGS__)
#endif

/**
 * Logs the message if the current log level >= to the DEBUG level.
 *
 * @note Doesn't construct the message if the current level is less than the DEBUG level
 */
#ifndef STREAM_LOG_DEBUG
#define STREAM_LOG_DEBUG(...) STREAM_LOG_CALL(log_level::debug, __VA_ARGS__)
#endif

/**
 * Logs the message if the current log level >= to the trace level.
 *
 * @note Doesn't construct the message if the current level is less than the trace level
 */
#ifndef STREAM_LOG_TRACE
#define STREAM_LOG_TRACE(...) STREAM_LOG_CALL(log_level::trace, __VA_ARGS__)
#endif

namespace stream_client {

// Forward declaration
class log_interface;

namespace detail {

/// Get/set global logger instance.
inline std::shared_ptr<log_interface> logger_instance(std::shared_ptr<log_interface> new_logger = nullptr);

} // namespace detail

/**
 * Base enum of log levels
 */
enum class log_level : int
{
    mute = -1,
    error = 0,
    warning,
    info,
    debug,
    trace,
};

/**
 * Type of the function used to log messages by the library logger.
 *
 * @param[in] level Level of the message.
 * @param[in] location Location of the message. Used for filtering messages from different parts of the library.
 * @param[in] message The body and main reason of message.
 *
 * @note  Thread-safe.
 */
using log_func_type = std::function<void(log_level level, const std::string& location, const std::string& message)>;

/**
 * Logger interface used by the library.
 */
class log_interface
{
public:
    /// Destructor.
    virtual ~log_interface() = default;

    /**
     * Set logger level.
     *
     * @param[in] level Level to setup.
     */
    virtual void set_level(log_level level) noexcept = 0;

    /**
     * Get logger level.
     *
     * @returns Actual logger level.
     */
    virtual log_level get_level() const noexcept = 0;

    /**
     * Log @p message produced from @p location with @p level.
     *
     * @param[in] level Level of the message.
     * @param[in] location Location of the message. Used to filter messages from different parts of the library.
     * @param[in] message The body of the message.
     *
     * @note Thread-safe.
     */
    virtual void message(log_level level, const std::string& location, const std::string& message) const noexcept = 0;
};

/// Logger calls passed callback function to log messages.
class func_logger: public stream_client::log_interface
{
public:
    func_logger(log_level level, stream_client::log_func_type log_func);

    /// Copy constructor.
    func_logger(const func_logger& other) = default;
    /// Copy assignment.
    func_logger& operator=(const func_logger& other) = default;
    /// Move constructor.
    func_logger(func_logger&& other) = default;
    /// Move assignment.
    func_logger& operator=(func_logger&& other) = default;

    /// Destructor.
    virtual ~func_logger() = default;

    /// Set logger level.
    virtual void set_level(log_level level) noexcept override;

    /// Get logger level.
    virtual log_level get_level() const noexcept override;

    /// Log arbitrary message.
    virtual void message(log_level level, const std::string& location, const std::string& message) const
        noexcept override;

private:
    log_level level_; ///< Logger level.
    stream_client::log_func_type log_func_; ///< Log function to call.
};

/// Logger print messages to std::cout with time and level.
class cout_logger: public stream_client::func_logger
{
public:
    static constexpr const char* kLevelPrefixes[] = {
        "ERROR", "WARNING", "INFO", "DEBUG", "trace",
    };

    /// Construct std::cout logger with @p level.
    cout_logger(log_level level = log_level::trace);

    /// Copy constructor.
    cout_logger(const cout_logger& other) = delete;
    /// Copy assignment.
    cout_logger& operator=(const cout_logger& other) = delete;
    /// Move constructor.
    cout_logger(cout_logger&& other) = delete;
    /// Move assignment.
    cout_logger& operator=(cout_logger&& other) = delete;

    /// Destructor.
    virtual ~cout_logger() = default;

private:
    void log(log_level level, const std::string& location, const std::string& message) const noexcept;

    mutable std::mutex mutex_; ///< Mutex to sync std::cout calls.
};

/**
 * Set logger for the library.
 * 
 * @note All set_logger variants overwrite each other. You can either set it via callback or an instance.
 *
 * @param[in] logger Logger instance to use. If nullptr logger will print messages to std::cout up to trace level.
 */
inline void set_logger(std::shared_ptr<log_interface> logger);

/**
 * Set log callback for the library.
 *
 * @note All set_logger variants overwrite each other. You can either set it via callback or an instance.
 *
 * @param[in] level Logger level to use.
 * @param[in] log_func Logger function to use.
 */
inline void set_logger(log_level level, log_func_type log_func);

/**
 * Set logger level.
 *
 * @param[in] level Level to setup.
 */
inline void set_log_level(log_level level) noexcept;

/**
 * Get logger level.
 *
 * @returns Actual logger level or log_level::MUTE if the current logger is nullptr.
 */
inline log_level get_log_level() noexcept;

/**
 * Log @p message produced from @p location with @p level.
 *
 * @param[in] level Level of the message.
 * @param[in] location Location of the message.
 * @param[in] message The body of the message.
 *
 * @note Thread-safe.
 */
inline void log_message(log_level level, const std::string& location, const std::string& message) noexcept;

} // namespace stream_client

#include "impl/logger.ipp"
```

Reply:
I'll drop the proposed `logger.ipp` for reference

```cpp
#pragma once

#include <iomanip>
#include <iostream>
#include <mutex>
#include <sstream>

namespace stream_client {
namespace detail {

/// Get/set global logger instance.
inline std::shared_ptr<log_interface> logger_instance(std::shared_ptr<log_interface> new_logger)
{
    // By default use cout logger with trace level
    static std::shared_ptr<log_interface> glob_instance = std::make_shared<cout_logger>();

    if (new_logger) {
        glob_instance = new_logger;
    }
    return glob_instance;
}

} // namespace detail

func_logger::func_logger(log_level level, stream_client::log_func_type log_func)
    : level_(level)
    , log_func_(log_func)
{
}

void func_logger::set_level(log_level level) noexcept
{
    level_ = level;
}

log_level func_logger::get_level() const noexcept
{
    return level_;
}

void func_logger::message(log_level level, const std::string& location, const std::string& message) const noexcept
{
    log_func_(level, location, message);
}

cout_logger::cout_logger(log_level level)
    : func_logger(level,
                  [this](log_level ll, const std::string& lc, const std::string& msg) { this->log(ll, lc, msg); })
{
}

void cout_logger::log(log_level level, const std::string& location, const std::string& message) const noexcept
{
    if (level > this->get_level()) {
        return;
    }

    std::stringstream ss;
    const auto now = std::chrono::system_clock::now();
    const std::time_t t_c = std::chrono::system_clock::to_time_t(now);
    ss << std::put_time(std::localtime(&t_c), "%Y-%m-%dT%H:%M:%SZ") << ": " << kLevelPrefixes[static_cast<int>(level)]
       << ": " << location << ": " << message << std::endl;

    std::lock_guard<std::mutex> lock(mutex_);
    std::cout << ss.str();
}

inline void set_logger(std::shared_ptr<log_interface> logger)
{
    detail::logger_instance(std::move(logger));
}

inline void set_logger(log_level level, log_func_type log_func)
{
    detail::logger_instance(std::make_shared<func_logger>(level, std::move(log_func)));
}

inline void set_log_level(log_level level) noexcept
{
    const auto logger = detail::logger_instance();
    if (logger) {
        logger->set_level(level);
    }
}

inline log_level get_log_level() noexcept
{
    const auto logger = detail::logger_instance();
    if (logger) {
        return logger->get_level();
    }
    return log_level::mute;
}

inline void log_message(log_level level, const std::string& location, const std::string& message) noexcept
{
    const auto logger = detail::logger_instance();
    if (logger) {
        logger->message(level, location, message);
    }
}

} // namespace stream_client
```